### DISCUSSION TITLE
ForeignKeyConstraint does not create a constraint in PostgreSQL only in SQLAlchemy ?

### DISCUSSION THREAD

**Discussion by chogovadze** [sqlalchemy/sqlalchemy#8501] [Usage Questions] [answered]
Hello everyone!

I noticed that when I create foreign key constraints using `ForeignKeyConstraint` in `__table_args__` the actual PostgreSQL database would be unaware of them. For example, when interacting through FastAPI using SQLAlchemy and the `ForeignKeyConstraint` is violated then the transaction would fail. However, if the same transaction is made directly to the DB then it works.

This behaviour is unlike `ForeignKey` which actually adds the constraint to the DB. I was hoping to have `ForeignKeyConstraints` that would have the `onupdate='CASCADE'` parameters but it does not seem clear to me how this would work. I am specifically using `ForeignKeyConstraint` instead of `ForeignKey` as I have composite foreign keys. 

Here is an example:

```python
import sqlalchemy.ext.declarative as _declarative

Base = _declarative.declarative_base()


class Parent(Base):
    __tablename__ = "parent"
    parent_attribute_1 = sql.Column(sql.String, primary_key=True)
    parent_attribute_2 = sql.Column(sql.String, primary_key=True)
    __table_args__ = ("schema": "my_schema"})



class Child(Base):
    __tablename__ = "child"
    parent_attribute_1 = sql.Column(sql.String, primary_key=True)
    parent_attribute_2 = sql.Column(sql.String, primary_key=True)
    child_attribute_1 = sql.Column(sql.String, primary_key=True)
    child_attribue_2 = sql.Column(sql.String, primary_key=True)
    __table_args__ = (
        sql.ForeignKeyConstraint(
                            [parent_attribute_1, parent_attribute_1], 
                            ["parent.parent_attribute_1", "parent.parent_attribute_2"], 
                            ondelete='CASCADE', onupdate='CASCADE'),
                                    {"schema": "my_schema"},
    )
```

Thank you for your time!

**Comment by zzzeek** [comment, marked as answer]
ForeignKey and ForeignKeyConstraint work in exactly the same way and ForeignKey actually generates the ForeignKeyConstraint, which is what does the work.

so your observation that "This behaviour is unlike ForeignKey which actually adds the constraint to the DB. " is not accurate from an expected behavior point of view - both should do the same thing , when SQLAlchemy emits CREATE TABLE statements.   

It seems likely that what you are observing has to do with the tables already existing when you call metadata.create_all(engine) or something like that, or perhaps the ForeignKeyConstraint specification is not getting into the mapping correctly.

Running your code the first error I get is:

    sqlalchemy.exc.ArgumentError: ForeignKeyConstraint with duplicate source column references are not supported.

fix that:

```
        ForeignKeyConstraint(
            [parent_attribute_1, parent_attribute_2],
            ["parent.parent_attribute_1", "parent.parent_attribute_2"],
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
```

next error:

    sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'child.parent_attribute_1' could not find table 'parent' with which to generate a foreign key to target column 'parent_attribute_1'

you have a schema there, that needs to be in the FK destination:

```py
        ForeignKeyConstraint(
            [parent_attribute_1, parent_attribute_2],
            ["test_schema.parent.parent_attribute_1", "test_schema.parent.parent_attribute_2"],
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
```

now it runs, generates FK

```sql
CREATE TABLE test_schema.parent (
	parent_attribute_1 VARCHAR NOT NULL, 
	parent_attribute_2 VARCHAR NOT NULL, 
	PRIMARY KEY (parent_attribute_1, parent_attribute_2)
)


2022-09-08 08:47:58,931 INFO sqlalchemy.engine.Engine [no key 0.00006s] {}
2022-09-08 08:47:58,955 INFO sqlalchemy.engine.Engine 
CREATE TABLE test_schema.child (
	parent_attribute_1 VARCHAR NOT NULL, 
	parent_attribute_2 VARCHAR NOT NULL, 
	child_attribute_1 VARCHAR NOT NULL, 
	child_attribue_2 VARCHAR NOT NULL, 
	PRIMARY KEY (parent_attribute_1, parent_attribute_2, child_attribute_1, child_attribue_2), 
	FOREIGN KEY(parent_attribute_1, parent_attribute_2) REFERENCES test_schema.parent (parent_attribute_1, parent_attribute_2) ON DELETE CASCADE ON UPDATE CASCADE
)
```

full script

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKeyConstraint
from sqlalchemy import String
import sqlalchemy.ext.declarative as _declarative

Base = _declarative.declarative_base()


class Parent(Base):
    __tablename__ = "parent"
    parent_attribute_1 = Column(String, primary_key=True)
    parent_attribute_2 = Column(String, primary_key=True)
    __table_args__ = {"schema": "test_schema"}


class Child(Base):
    __tablename__ = "child"
    parent_attribute_1 = Column(String, primary_key=True)
    parent_attribute_2 = Column(String, primary_key=True)
    child_attribute_1 = Column(String, primary_key=True)
    child_attribue_2 = Column(String, primary_key=True)
    __table_args__ = (
        ForeignKeyConstraint(
            [parent_attribute_1, parent_attribute_2],
            [
                "test_schema.parent.parent_attribute_1",
                "test_schema.parent.parent_attribute_2",
            ],
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
        {"schema": "test_schema"},
    )


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
```

**Reply by chogovadze** [reply]
Thank you very much for your detailed answer and sorry for the broken example. Is there any way to force the constraints to be updated to include the `CASCADE` behaviour if the table has already been created without them and contains data ?

**Reply by zzzeek** [reply]
ah that's a bigger issue, but yes you have to DROP and CREATE the constraint correctly.    There's a few levels of how you can do this, depends on if this is just a one-off or some kind of pattern you'd need to repeat.  

if just a one off for a database that you can't just recreate, I'd write a simple SQL script and just run it.  here's a blog that runs through this specific case: https://codeburst.io/edit-existing-foreign-key-constraints-in-postgresql-with-the-alter-table-command-c1de89256ff7

if this is for a database that you can recreate, then just do that.

if this is something you need to automate into your application so that you can add new migrations to the schema on an ongoing basis, look into using [alembic](https://alembic.sqlalchemy.org/en/latest/).

**Reply by chogovadze** [reply]
I see! Thank you for the pointers. I would need to change quite a few constraints and wished to do this without having to write a SQL query for each of them

**Reply by zzzeek** [reply]
you can automate it in a lot of ways.   you could run a pg dump into a SQL script, alter the CREATE statements in the script, then run it into a new database.     You can use SQLAlchemy and/or alembic to generate DROP / CREATE statements also.  like to emit for your table:

```py
from sqlalchemy.schema import AddConstraint

constraint = list(Child.__table__.foreign_key_constraints)[0]

print(AddConstraint(constraint))
```

```
ALTER TABLE test_schema.child ADD FOREIGN KEY(parent_attribute_1, parent_attribute_2) REFERENCES test_schema.parent (parent_attribute_1, parent_attribute_2) ON DELETE CASCADE ON UPDATE CASCADE
```

however, to DROP it first you need to know its name.   If you didn't give them explicit names, PG has generated these for you.  you can see these names using inspect

```py
>>> from sqlalchemy import inspect
>>> inspect(e).get_foreign_keys("child", "test_schema")
[{'name': 'child_parent_attribute_1_parent_attribute_2_fkey', 'constrained_columns': ['parent_attribute_1', 'parent_attribute_2'], 'referred_schema': 'test_schema', 'referred_table': 'parent', 'referred_columns': ['parent_attribute_1', 'parent_attribute_2'], 'options': {'onupdate': 'CASCADE', 'ondelete': 'CASCADE'}}]
```

so the DROP for that one would look like:

```py
>>> constraint.name = 'child_parent_attribute_1_parent_attribute_2_fkey' 
>>> from sqlalchemy.schema import DropConstraint
>>> print(DropConstraint(constraint))
ALTER TABLE test_schema.child DROP CONSTRAINT child_parent_attribute_1_parent_attribute_2_fkey
```

the whole thing we're doing with SQLAlchemy is writing Python code to manipulate the database so...this is the real deal!

**Reply by chogovadze** [reply]
Amazing! Thank you so much @zzzeek!
